service/controller/app/v1/resource/cordonchart/resource.go

```diff
@@ -0,0 +1,135 @@
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/giantswarm/app-operator/pkg/annotation"
+	"github.com/giantswarm/app-operator/service/controller/app/v1/key"
+)
+
+const (
+	// Name is the identifier of the resource.
+	Name = "cordonchartv1"
+)
+
+//Config represents the configuration used to create a new cordonchart resource.
+type Config struct {
+	// Dependencies.
+	Logger micrologger.Logger
+
+	// Settings.
+	ChartNamespace string
+}
+
+type patchSpec struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value,omitempty"`
+}
+
+// Resource implements the cordonchart resource.
+type Resource struct {
+	// Dependencies.
+	logger micrologger.Logger
+
+	// Settings.
+	chartNamespace string
+}
+
+// New creates a new configured cordonchart resource.
+func New(config Config) (*Resource, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.ChartNamespace == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ChartNamespace must not be empty", config)
+	}
+
+	r := &Resource{
+		logger:         config.Logger,
+		chartNamespace: config.ChartNamespace,
+	}
+
+	return r, nil
+}
+
+func (r Resource) Name() string {
+	return Name
+}
+
+func (r *Resource) addCordon(ctx context.Context, cr v1alpha1.App, client versioned.Interface) error {
+	var err error
+	var patchByte []byte
+	{
+		patch := []patchSpec{
+			{
+				Op:   "add",
+				Path: "/metadata/annotations",
+				Value: map[string]string{
+					replacePrefix(annotation.CordonReason): key.CordonReason(cr),
+					replacePrefix(annotation.CordonUntil):  key.CordonUntil(cr),
+				},
+			},
+		}
+
+		patchByte, err = json.Marshal(patch)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	_, err = client.ApplicationV1alpha1().Charts(r.chartNamespace).Patch(cr.GetName(), types.JSONPatchType, patchByte)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Resource) deleteCordon(ctx context.Context, cr v1alpha1.App, client versioned.Interface) error {
+	var err error
+	var patchByte []byte
+
+	{
+		patch := []patchSpec{
+			{
+				Op:   "remove",
+				Path: replaceToEscape(fmt.Sprintf("/metadata/annotations/%s", replacePrefix(annotation.CordonUntil))),
+			},
+			{
+				Op:   "remove",
+				Path: replaceToEscape(fmt.Sprintf("/metadata/annotations/%s", replacePrefix(annotation.CordonReason))),
+			},
+		}
+
+		patchByte, err = json.Marshal(patch)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	_, err = client.ApplicationV1alpha1().Charts(r.chartNamespace).Patch(cr.GetName(), types.JSONPatchType, patchByte)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func replacePrefix(from string) string {
+	return strings.Replace(from, "app-operator.giantswarm.io", "chart-operator.giantswarm.io", 1)
+}
+
+func replaceToEscape(from string) string {
+	return strings.Replace(from, "/", "~1", -1)
```

Comment:
All `/` characters should change to `~1` according to JSON Merge RFC documentations. https://tools.ietf.org/html/rfc6902#appendix-A.14

Reply:
Why do we need that though? In the `statusresource` and I am sure in other places this works without this replacement. See https://github.com/giantswarm/statusresource/blob/master/resource.go.

Reply:
if key name contains `/` or `~`, We should convert it to escape characters otherwise the key name `chart-operator.giantswarm.io/cordon-until` would have two depth even though it is one depth in real. So far I don't see the same case in `statusresource`

Reply:
If I understand the code correctly what happens here is that the patch path `/metadata/annotations/...` becomes `~1metadata~1annotations~1...` and this looks pretty wrong to me. Did you test this? What is the use case for this weird escaping here? `/` is a path separator AFAIK which is exactly what happens in the `statusresource`.

Reply:
After I changed the code for adding `annotation`, I didn't test it enough for this.
That's wrong I should only change the last key name which contains `/` so
`/metadata/annotations/"chart-operator.giantswarm.io/cordon-until"` should change into 
`/metadata/annotations/"chart-operator.giantswarm.io~1cordon-until"`

Reply:
That makes sense yes.